fl4health/parameter_exchange/layer_exchanger.py

```diff
@@ -92,88 +91,76 @@ def pull_parameters(self, parameters: NDArrays, model: nn.Module, config: Option
+
+    def set_layer_selection_function(
+        self, layer_selection_function: Callable[..., Tuple[NDArrays, List[str]]]
+    ) -> None:
+        self.layer_selection_function = layer_selection_function
 
     def set_filter_mode(self, filter_by_percentage: bool) -> None:
         self.filter_by_percentage = filter_by_percentage
 
     def set_normalization_mode(self, normalize: bool) -> None:
         self.normalize = normalize
 
+    def set_select_drift_more(self, select_drift_more: bool) -> None:
+        self.select_drift_more = select_drift_more
+
     def _calculate_drift_norm(self, t1: torch.Tensor, t2: torch.Tensor) -> float:
         t_diff = (t1 - t2).float()
         drift_norm = torch.linalg.norm(t_diff)
         if self.normalize:
             drift_norm /= torch.numel(t_diff)
         return drift_norm.item()
 
-    def filter_layers_by_threshold(self, model: nn.Module, initial_model: nn.Module) -> Tuple[NDArrays, List[str]]:
-        """
-        Return those layers of model that deviate (in l2 norm) away from corresponding layers of
-        self.initial_model by at least self.threshold.
-        """
-        layer_names = []
-        layers_to_transfer = []
-        initial_model_states = initial_model.state_dict()
-        model_states = model.state_dict()
-        for layer_name, layer_param in model_states.items():
-            layer_param_past = initial_model_states[layer_name]
-            drift_norm = self._calculate_drift_norm(layer_param, layer_param_past)
-            if drift_norm >= self.threshold:
-                layers_to_transfer.append(layer_param.cpu().numpy())
-                layer_names.append(layer_name)
-        return layers_to_transfer, layer_names
-
     def set_threshold(self, new_threshold: float) -> None:
-        self.threshold = new_threshold
+        self.norm_threshold = new_threshold
 
     def set_percentage(self, new_percentage: float) -> None:
         assert 0 < new_percentage <= 1
         self.exchange_percentage = new_percentage
 
-    def filter_layers_by_percentage(self, model: nn.Module, initial_model: nn.Module) -> Tuple[NDArrays, List[str]]:
-        names_to_norm_drift = {}
-        initial_model_states = initial_model.state_dict()
-        model_states = model.state_dict()
-
-        for layer_name, layer_param in model_states.items():
-            layer_param_past = initial_model_states[layer_name]
-            drift_norm = self._calculate_drift_norm(layer_param, layer_param_past)
-            names_to_norm_drift[layer_name] = drift_norm
-
-        total_param_num = len(names_to_norm_drift.keys())
-        num_param_exchange = int(math.ceil(total_param_num * self.exchange_percentage))
-        param_to_exchange_names = sorted(names_to_norm_drift.keys(), key=lambda x: names_to_norm_drift[x])[
-            :(num_param_exchange)
-        ]
-        return [model_states[name].cpu().numpy() for name in param_to_exchange_names], param_to_exchange_names
+    def select_parameters(
```

Comment:
Since we have two different select parameters functionality (filter_by_percentage and filter_by_threshold) is it better to just have separate classes? Perhaps you can extend PartialWeightExchanger to add functionality that is common between these two drift based exchangers (ie push_parameters and pull_parameters) and then extend that class and implement the select_parameters method?

The only reason I suggest this is because its a bit awkward the fact you can potentially pass both a norm_threshold and an exchange_percentage even though you only use one and then you have to pass an extra bool (filter_by_percentage) to clarify. 

I could be completely missing something so feel free to fill me in :)

Reply:
An alternative to this is that the selection function that is provided needs to be "pre-constructed" using partial with all but `model` and `initial_model`. That is, it only has a signature of `Callable[[nn.Module, nn.Module], Tuple[NDArrays, List[str]]]`. Then the `parameter_selection_criteria.py` file can have constructors based on the auxiliary inputs. We would then just have a `filter_by_percentage` and `filter_by_threshold` function constructor.

Reply:
@emersodb Yes, I like that! We don't have to impose as strong of constraint on the type signature through this "pre-constructing". It worked out nicely in this case as you [asserted above](https://github.com/VectorInstitute/FL4Health/pull/103/files#r1476423922) but its nice to have more flexibility if needed